File: src/main/java/pl/dmichalski/algorithms/_6_max_subarray_sum/MaxSubarraySumService2.java
package pl.dmichalski.algorithms._6_max_subarray_sum;

class MaxSubarraySumService2 {

    /**
     * O(n) complexity
     */
    public Integer maxSubarraySum(int[] values, int numbersCountToCountSum) {
        if (values == null || values.length < numbersCountToCountSum) {
            return null;
        }

        int maxSum = 0;
        for (int i = 0; i < numbersCountToCountSum; i++) {
            maxSum += values[i];
        }

        int tmpSum = maxSum;
        for (int i = numbersCountToCountSum; i < values.length; i++) {
            tmpSum = tmpSum - values[i-numbersCountToCountSum] + values[i];
            maxSum = Math.max(maxSum, tmpSum);
        }

        return maxSum;
    }
}

File: src/main/java/pl/dmichalski/algorithms/_6_max_subarray_sum/MaxSubarraySumService2.kt
package pl.dmichalski.algorithms._6_max_subarray_sum

internal class MaxSubarraySumService2 {

    /**
     * O(n) complexity
     */
    fun maxSubarraySum(values: IntArray?, numbersCountToCountSum: Int): Int? {
        if (values == null || values.size < numbersCountToCountSum) {
            return null
        }

        var maxSum = 0
        for (i in 0 until numbersCountToCountSum) {
            maxSum += values[i]
        }
        var tmpSum = maxSum
        for (i in numbersCountToCountSum until values.size) {
            tmpSum = tmpSum - values[i - numbersCountToCountSum] + values[i]
            maxSum = Math.max(maxSum, tmpSum)
        }

        return maxSum
    }

}
